File: python/gamex/platform.py
from __future__ import annotations
import os
from enum import Enum
from gamex.pak import PakFile
from openstk.gfx.gfx import IObjectManager, IMaterialManager, IShaderManager, ITextureManager, PlatformStats
from openstk.gfx.gfx_render import IMaterial
from openstk.gfx.gfx_texture import ITexture
from openstk.sfx.sfx import IAudioManager

# IFileSystem
class IFileSystem:
    def glob(self, path: str, searchPattern: str) -> list[str]: pass
    def fileExists(self, path: str) -> bool: pass
    def fileInfo(self, path: str) -> (str, int): pass
    def openReader(self, path: str, mode: str = 'rb') -> Reader: pass
    def findPaths(self, path: str, searchPattern: str) -> str:
        if (expandStartIdx := searchPattern.find('(')) != -1 and \
            (expandMidIdx := searchPattern.find(':', expandStartIdx)) != -1 and \
            (expandEndIdx := searchPattern.find(')', expandMidIdx)) != -1 and \
            expandStartIdx < expandEndIdx:
            for expand in searchPattern[expandStartIdx + 1: expandEndIdx].split(':'):
                for found in self.findPaths(path, searchPattern[:expandStartIdx] + expand + searchPattern[expandEndIdx+1:]): yield found
            return
        for path in self.glob(path, searchPattern): yield path

# AudioBuilderBase
class AudioBuilderBase:
    def createAudio(self, path: object) -> Audio: pass
    def deleteAudio(self, audio: Audio) -> None: pass

# AudioManager
class AudioManager(IAudioManager):
    _pakFile: PakFile
    _builder: AudioBuilderBase
    _cachedAudios: dict[object, (Audio, object)] = {}
    _preloadTasks: dict[object, object] = {}

    def __init__(self, pakFile: PakFile, builder: AudioBuilderBase):
        self._pakFile = pakFile
        self._builder = builder

    def createAudio(self, key: object) -> (Audio, object):
        if path in self._cachedAudios: return self._cachedAudios[path]
        # load & cache the audio.
        tag = self._loadAudio(path)
        audio = self._builder.createAudio(tag) if tag else None
        self._cachedAudios[path] = (audio, tag)
        return (audio, tag)

    def preloadAudio(self, path: object) -> None:
        if path in self._cachedAudios: return
        # start loading the audio file asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(object, path)

    def deleteAudio(self, path: object) -> None:
        if not path in self._cachedAudios: return
        self._builder.deleteAudio(self._cachedAudios[0])
        self._cachedAudios.remove(path)

    async def _loadAudio(self, path: object) -> ITexture:
        assert(not path in self._cachedAudios)
        self.preloadAudio(s)
        source = await self._preloadTasks[path]
        self._preloadTasks.remove(path)
        return source

# TextureBuilderBase
class TextureBuilderBase:
    maxTextureMaxAnisotropy: int = PlatformStats.maxTextureMaxAnisotropy
    defaultTexture: Texture
    def createTexture(self, reuse: Texture, source: ITexture, level: range = None) -> Texture: pass
    def createSolidTexture(self, width: int, height: int, rgba: list[float]) -> Texture: pass
    def createNormalMap(self, texture: Texture, strength: float) -> Texture: pass
    def deleteTexture(self, texture: Texture) -> None: pass

# TextureManager
class TextureManager(ITextureManager):
    _pakFile: PakFile
    _builder: TextureBuilderBase
    _cachedTextures: dict[object, (Texture, object)] = {}
    _preloadTasks: dict[object, ITexture] = {}

    def __init__(self, pakFile: PakFile, builder: TextureBuilderBase):
        self._pakFile = pakFile
        self._builder = builder

    def createSolidTexture(self, width: int, height: int, rgba: list[float] = None) -> Texture: return self._builder.createSolidTexture(width, height, rgba)

    def createNormalMap(self, source: Texture, strength: float) -> Texture: return self._builder.createNormalMap(source, strength)

    @property
    def defaultTexture(self) -> Texture: return self._builder.defaultTexture

    def createTexture(self, path: object, level: range = None) -> (Texture, object):
        if path in self._cachedTextures: return self._cachedTextures[path]
        # load & cache the texture.
        tag = path if isinstance(path, ITexture) else self._loadTexture(path)
        texture = self._builder.createTexture(None, tag, level) if tag else self._builder.defaultTexture
        self._cachedTextures[path] = (texture, tag)
        return (texture, tag)

    def reloadTexture(self, path: object, level: range = None) -> (Texture, object):
        if path not in self._cachedTextures: return (None, None)
        c = self._cachedTextures[path]
        self._builder.createTexture(c[0], c[1], level)
        return c

    def preloadTexture(self, path: object) -> None:
        if path in self._cachedTextures: return
        # start loading the texture file asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(object, path)

    def deleteTexture(self, path: object) -> None:
        if not path in self._cachedTextures: return
        self._builder.deleteTexture(self._cachedTextures[0])
        self._cachedTextures.remove(path)

    async def _loadTexture(self, path: object) -> ITexture:
        assert(not path in self._cachedTextures)
        self.preloadTexture(s)
        source = await self._preloadTasks[path]
        self._preloadTasks.remove(path)
        return source

# ShaderBuilderBase
class ShaderBuilderBase:
    def createShader(self, path: object, args: dict[str, bool]) -> Shader: pass

# ShaderManager
class ShaderManager(IShaderManager):
    emptyArgs: dict[str, bool] = {}
    _pakFile: PakFile
    _builder: ShaderBuilderBase

    def __init__(self, pakFile: PakFile, builder: ShaderBuilderBase):
        self._pakFile = pakFile
        self._builder = builder
    
    def createShader(self, path: object, args: dict[str, bool] = None) -> (Shader, object):
        return (self._builder.createShader(path, args or self.emptyArgs), None)

# ObjectBuilderBase
class ObjectBuilderBase:
    def ensurePrefab(self) -> None: pass
    def createNewObject(self, prefab: object) -> object: pass
    def createObject(self, source: object, materialManager: IMaterialManager) -> object: pass

# ObjectManager
class ObjectManager(IObjectManager):
    _pakFile: PakFile
    _materialManager: IMaterialManager
    _builder: ObjectBuilderBase
    _cachedObjects: dict[str, object] = {}
    _preloadTasks: dict[str, object] = {}

    def __init__(self, pakFile: PakFile, materialManager: IMaterialManager, builder: ObjectBuilderBase):
        self._pakFile = pakFile
        self._materialManager = materialManager
        self._builder = builder

    def createNewObject(self, path: object) -> (object, object):
        tag = None
        self._builder.ensurePrefab()
        # load & cache the prefab.
        if not path in self._cachedObjects: prefab = self._cachedObjects[path] = (self._loadObject(path), tag)
        else: prefab = self._cachedObjects[path]
        # instantiate the prefab.
        return (self._builder.createNewObject(prefab[0]), prefab[1])
 
    def preloadObject(self, path: object) -> None:
        if path in self._cachedPrefabs: return
        # start loading the object asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(object, path)

    async def _loadObject(self, path: object) -> object:
        assert(not path in self._cachedPrefabs)
        self.preloadObject(path)
        source = await self._preloadTasks[path]
        self._preloadTasks.remove(path)
        return self._builder.buildObject(source, self._materialManager)

# MaterialBuilderBase
class MaterialBuilderBase:
    textureManager : ITextureManager
    normalGeneratorIntensity: float = 0.75
    defaultMaterial: Material

    def __init__(self, textureManager: ITextureManager): self.TextureManager = textureManager
    def createMaterial(self, path: object) -> Material: pass

# MaterialManager
class MaterialManager(IMaterialManager):
    _pakFile: PakFile
    _builder: MaterialBuilderBase
    _cachedMaterials: dict[object, (Material, object)] = {}
    _preloadTasks: dict[object, IMaterial] = {}

    textureManager: ITextureManager

    def __init__(self, pakFile: PakFile, textureManager: ITextureManager, builder: MaterialBuilderBase):
        self._pakFile = pakFile
        self._textureManager = textureManager
        self._builder = builder

    def createMaterial(self, path: object) -> (Material, object):
        if path in self._cachedMaterials: return self._cachedMaterials[path]
        # load & cache the material.
        source = path if isinstance(path, IMaterial) else self._loadMaterial(path)
        material = self._builder.createMaterial(source) if source else self._builder.defaultMaterial
        tag = None #source.data if source else None
        self._cachedMaterials[path] = (material, tag)
        return (material, tag)

    def preloadMaterial(self, path: object) -> None:
        if path in self._cachedMaterials: return
        # start loading the material file asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(IMaterial, path)

    async def _loadMaterial(self, path: object) -> IMaterial:
        assert(not path in self._cachedMaterials)
        self.preloadMaterial(path)
        source = await self.preloadTasks[path]
        self.preloadTasks.remove(path)
        return source

# typedefs
# class Type(Enum): pass
# class OS(Enum): pass

# PlatformX
class Platform:
    enabled: bool = True
    id: str = None
    name: str = None
    tag: str = None
    gfxFactory: callable = None
    sfxFactory: callable = None
    logFunc: callable = lambda a: print(a)
    def __init__(self, id: str, name: str): self.id = id; self.name = name
    def activate(self) -> None: pass
    def deactivate(self) -> None: pass

# UnknownPlatform
class UnknownPlatform(Platform):
    def __init__(self): super().__init__('UK', 'Unknown')
UnknownPlatform.This = UnknownPlatform()

# PlatformX
class PlatformX:
    class OS(Enum):
        Windows = 1
        OSX = 2
        Linux = 3
        Android = 4

    @staticmethod
    def decodeOptions() -> object:
        path = f'{os.getenv("APPDATA")}.gamex'
        return None

    platformOS: OS = OS.Windows
    platforms: set[object] = {UnknownPlatform.This}
    inTestHost: bool = False
    applicationPath = os.getcwd()
    options = decodeOptions()
    current: Platform = None

    class Stats:
        maxTextureMaxAnisotropy: int

    # activate
    @staticmethod
    def activate(platform: Platform) -> None:
        if not platform or not platform.enabled: platform = UnknownPlatform.This
        PlatformX.platforms.add(platform)
        current = PlatformX.current
        if current != platform:
            if current: current.deactivate()
            if platform: platform.activate()
            PlatformX.current = platform
        return platform

    @staticmethod
    def createMatcher(searchPattern: str) -> callable:
        if not searchPattern: return lambda x: True
        wildcardCount = searchPattern.count('*')
        if wildcardCount <= 0: return lambda x: x.casefold() == searchPattern.casefold()
        elif wildcardCount == 1:
            newPattern = searchPattern.replace('*', '')
            if searchPattern.startswith('*'): return lambda x: x.casefold().endswith(newPattern)
            elif searchPattern.endswith('*'): return lambda x: x.casefold().startswith(newPattern)
        regexPattern = f'^{re.escape(searchPattern).replace('\\*', '.*')}$'
        def lambdaX(x: str):
            try: return re.match(x, regexPattern)
            except: return False
        return lambdaX

    @staticmethod
    def decodePath(path: str, rootPath: str = None) -> str:
        lowerPath = path.lower()
        return f'{rootPath}{path[6:]}' if lowerPath.startswith('%path%') else \
        f'{PlatformX.applicationPath}{path[9:]}' if lowerPath.startswith('%apppath%') else \
        f'{os.getenv("APPDATA")}{path[9:]}' if lowerPath.startswith('%appdata%') else \
        f'{os.getenv("LOCALAPPDATA")}{path[14:]}' if lowerPath.startswith('%localappdata%') else \
        path

# TestGfx
class TestGfx:
    def __init__(self, source): self._source = source

# TestGfx
class TestSfx:
    def __init__(self, source): self._source = source

# TestPlatform
class TestPlatform(Platform):
    def __init__(self):
        super().__init__('TT', 'Test')
        self.gfxFactory = staticmethod(lambda source: TestGfx(source))
        self.sfxFactory = staticmethod(lambda source: TestSfx(source))
TestPlatform.This = TestPlatform()

PlatformX.current = PlatformX.activate(TestPlatform.This if PlatformX.inTestHost else UnknownPlatform.This)

File: dotnet/Base/GameX/Platform.cs
using OpenStack.Gfx;
using OpenStack.Gfx.Textures;
using OpenStack.Sfx;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;
using static OpenStack.Debug;
using Stopwatch = System.Diagnostics.Stopwatch;

namespace GameX.Platforms;

#region IFileSystem

/// <summary>
/// IFileSystem
/// </summary>
public interface IFileSystem
{
    IEnumerable<string> Glob(string path, string searchPattern);
    bool FileExists(string path);
    (string path, long length) FileInfo(string path);
    BinaryReader OpenReader(string path);
    BinaryWriter OpenWriter(string path);
}

#endregion

#region AudioBuilderBase

public abstract class AudioBuilderBase<Audio>
{
    public abstract Audio CreateAudio(object path);
    public abstract void DeleteAudio(Audio audio);
}

#endregion

#region AudioManager

public class AudioManager<Audio>(PakFile pakFile, AudioBuilderBase<Audio> builder) : IAudioManager<Audio>
{
    readonly PakFile PakFile = pakFile;
    readonly AudioBuilderBase<Audio> Builder = builder;
    readonly Dictionary<object, (Audio aud, object tag)> CachedAudios = [];
    readonly Dictionary<object, Task<object>> PreloadTasks = [];

    public (Audio aud, object tag) CreateAudio(object path)
    {
        if (CachedAudios.TryGetValue(path, out var c)) return c;
        // load & cache the audio.
        var tag = LoadAudio(path).Result;
        var audio = tag != null ? Builder.CreateAudio(tag) : default;
        CachedAudios[path] = (audio, tag);
        return (audio, tag);
    }

    public void PreloadAudio(object path)
    {
        if (CachedAudios.ContainsKey(path)) return;
        // start loading the texture file asynchronously if we haven't already started.
        if (!PreloadTasks.ContainsKey(path)) PreloadTasks[path] = PakFile.LoadFileObject<object>(path);
    }

    public void DeleteAudio(object path)
    {
        if (!CachedAudios.TryGetValue(path, out var c)) return;
        Builder.DeleteAudio(c.aud);
        CachedAudios.Remove(path);
    }

    async Task<object> LoadAudio(object path)
    {
        Assert(!CachedAudios.ContainsKey(path));
        PreloadAudio(path);
        var source = await PreloadTasks[path];
        PreloadTasks.Remove(path);
        return source;
    }
}

#endregion

#region TextureBuilderBase

public abstract class TextureBuilderBase<Texture>
{
    public static int MaxTextureMaxAnisotropy
    {
        get => PlatformStats.MaxTextureMaxAnisotropy;
        set => PlatformStats.MaxTextureMaxAnisotropy = value;
    }

    public abstract Texture DefaultTexture { get; }
    public abstract Texture CreateTexture(Texture reuse, ITexture source, Range? level = null);
    public abstract Texture CreateSolidTexture(int width, int height, float[] rgba);
    public abstract Texture CreateNormalMap(Texture texture, float strength);
    public abstract void DeleteTexture(Texture texture);
}

#endregion

#region TextureManager

public class TextureManager<Texture>(PakFile pakFile, TextureBuilderBase<Texture> builder) : ITextureManager<Texture>
{
    readonly PakFile PakFile = pakFile;
    readonly TextureBuilderBase<Texture> Builder = builder;
    readonly Dictionary<object, (Texture tex, object tag)> CachedTextures = [];
    readonly Dictionary<object, Task<ITexture>> PreloadTasks = [];

    public Texture CreateSolidTexture(int width, int height, params float[] rgba) => Builder.CreateSolidTexture(width, height, rgba);

    public Texture CreateNormalMap(Texture source, float strength) => Builder.CreateNormalMap(source, strength);

    public Texture DefaultTexture => Builder.DefaultTexture;

    public (Texture tex, object tag) CreateTexture(object path, Range? level = null)
    {
        if (CachedTextures.TryGetValue(path, out var c)) return c;
        // load & cache the texture.
        var tag = path is ITexture z ? z : LoadTexture(path).Result;
        var texture = tag != null ? Builder.CreateTexture(default, tag, level) : Builder.DefaultTexture;
        CachedTextures[path] = (texture, tag);
        return (texture, tag);
    }

    public (Texture tex, object tag) ReloadTexture(object path, Range? level = null)
    {
        if (!CachedTextures.TryGetValue(path, out var c)) return (default, default);
        Builder.CreateTexture(c.tex, (ITexture)c.tag, level);
        return c;
    }

    public void PreloadTexture(object path)
    {
        if (CachedTextures.ContainsKey(path)) return;
        // start loading the texture file asynchronously if we haven't already started.
        if (!PreloadTasks.ContainsKey(path)) PreloadTasks[path] = PakFile.LoadFileObject<ITexture>(path);
    }

    public void DeleteTexture(object path)
    {
        if (!CachedTextures.TryGetValue(path, out var c)) return;
        Builder.DeleteTexture(c.tex);
        CachedTextures.Remove(path);
    }

    async Task<ITexture> LoadTexture(object path)
    {
        Assert(!CachedTextures.ContainsKey(path));
        PreloadTexture(path);
        var source = await PreloadTasks[path];
        PreloadTasks.Remove(path);
        return source;
    }
}

#endregion

#region ShaderBuilderBase

public abstract class ShaderBuilderBase<Shader>
{
    public abstract Shader CreateShader(object path, IDictionary<string, bool> args = null);
}

#endregion

#region ShaderManager

public class ShaderManager<Shader>(PakFile pakFile, ShaderBuilderBase<Shader> builder) : IShaderManager<Shader>
{
    static readonly Dictionary<string, bool> EmptyArgs = [];
    readonly PakFile PakFile = pakFile;
    readonly ShaderBuilderBase<Shader> Builder = builder;

    public (Shader sha, object tag) CreateShader(object path, IDictionary<string, bool> args = null)
        => (Builder.CreateShader(path, args ?? EmptyArgs), null);
}

#endregion

#region ObjectBuilderBase

public abstract class ObjectBuilderBase<Object, Material, Texture>
{
    public abstract void EnsurePrefab();
    public abstract Object CreateNewObject(Object prefab);
    public abstract Object CreateObject(object source, IMaterialManager<Material, Texture> materialManager);
}

#endregion

#region ObjectManager

public class ObjectManager<Object, Material, Texture>(PakFile pakFile, IMaterialManager<Material, Texture> materialManager, ObjectBuilderBase<Object, Material, Texture> builder) : IObjectManager<Object, Material, Texture>
{
    readonly PakFile PakFile = pakFile;
    readonly IMaterialManager<Material, Texture> MaterialManager = materialManager;
    readonly ObjectBuilderBase<Object, Material, Texture> Builder = builder;
    readonly Dictionary<object, (Object obj, object tag)> CachedObjects = new Dictionary<object, (Object obj, object tag)>();
    readonly Dictionary<object, Task<object>> PreloadTasks = new Dictionary<object, Task<object>>();

    public (Object obj, object tag) CreateObject(object path)
    {
        Builder.EnsurePrefab();
        // load & cache the prefab.
        if (!CachedObjects.TryGetValue(path, out var prefab)) prefab = CachedObjects[path] = LoadObject(path).Result;
        return (Builder.CreateNewObject(prefab.obj), prefab.tag);
    }

    public void PreloadObject(object path)
    {
        if (CachedObjects.ContainsKey(path)) return;
        // start loading the object asynchronously if we haven't already started.
        if (!PreloadTasks.ContainsKey(path)) PreloadTasks[path] = PakFile.LoadFileObject<object>(path);
    }

    async Task<(Object obj, object tag)> LoadObject(object path)
    {
        Assert(!CachedObjects.ContainsKey(path));
        PreloadObject(path);
        var source = await PreloadTasks[path];
        PreloadTasks.Remove(path);
        return (Builder.CreateObject(source, MaterialManager), source);
    }
}

#endregion

#region MaterialBuilderBase

public abstract class MaterialBuilderBase<Material, Texture>(ITextureManager<Texture> textureManager)
{
    protected ITextureManager<Texture> TextureManager = textureManager;
    public float? NormalGeneratorIntensity = 0.75f;
    public abstract Material DefaultMaterial { get; }

    public abstract Material CreateMaterial(object path);
}

#endregion

#region MaterialManager

/// <summary>
/// Manages loading and instantiation of materials.
/// </summary>
public class MaterialManager<Material, Texture>(PakFile pakFile, ITextureManager<Texture> textureManager, MaterialBuilderBase<Material, Texture> builder) : IMaterialManager<Material, Texture>
{
    readonly PakFile PakFile = pakFile;
    readonly MaterialBuilderBase<Material, Texture> Builder = builder;
    readonly Dictionary<object, (Material material, object tag)> CachedMaterials = [];
    readonly Dictionary<object, Task<IMaterial>> PreloadTasks = [];
    public ITextureManager<Texture> TextureManager { get; } = textureManager;

    public (Material mat, object tag) CreateMaterial(object path)
    {
        if (CachedMaterials.TryGetValue(path, out var c)) return c;
        // load & cache the material.
        var source = path is IMaterial z ? z : LoadMaterial(path).Result;
        var material = source != null ? Builder.CreateMaterial(source) : Builder.DefaultMaterial;
        object tag = null; // source?.Data;
        CachedMaterials[path] = (material, tag);
        return (material, tag);
    }

    public void PreloadMaterial(object path)
    {
        if (CachedMaterials.ContainsKey(path)) return;
        // start loading the material file asynchronously if we haven't already started.
        if (!PreloadTasks.ContainsKey(path)) PreloadTasks[path] = PakFile.LoadFileObject<IMaterial>(path);
    }

    async Task<IMaterial> LoadMaterial(object path)
    {
        Assert(!CachedMaterials.ContainsKey(path));
        PreloadMaterial(path);
        var source = await PreloadTasks[path];
        PreloadTasks.Remove(path);
        return source;
    }
}

#endregion

#region Platform

/// <summary>
/// Gets the platform.
/// </summary>
public abstract class Platform(string id, string name)
{
    /// <summary>
    /// Gets the active.
    /// </summary>
    public bool Enabled = true;

    /// <summary>
    /// Gets the platform id.
    /// </summary>
    public readonly string Id = id;

    /// <summary>
    /// Gets the platform name.
    /// </summary>
    public readonly string Name = name;

    /// <summary>
    /// Gets the platform tag.
    /// </summary>
    public string Tag;

    /// <summary>
    /// Gets the platforms gfx factory.
    /// </summary>
    public Func<PakFile, IOpenGfx> GfxFactory = source => null; // throw new Exception("No GfxFactory");

    /// <summary>
    /// Gets the platforms sfx factory.
    /// </summary>
    public Func<PakFile, IOpenSfx> SfxFactory = source => null; // throw new Exception("No SfxFactory");

    /// <summary>
    /// Gets the platforms assert func.
    /// </summary>
    public Action<bool> AssertFunc = x => System.Diagnostics.Debug.Assert(x);

    /// <summary>
    /// Gets the platforms log func.
    /// </summary>
    public Action<string> LogFunc = a => System.Diagnostics.Debug.Print(a);

    /// <summary>
    /// Gets the platforms logformat func.
    /// </summary>
    public Action<string, object[]> LogFormatFunc = (a, b) => System.Diagnostics.Debug.Print(a, b);

    /// <summary>
    /// Activates the platform.
    /// </summary>
    public virtual void Activate()
    {
        OpenStack.Debug.AssertFunc = AssertFunc;
        OpenStack.Debug.LogFunc = LogFunc;
        OpenStack.Debug.LogFormatFunc = LogFormatFunc;
    }

    /// <summary>
    /// Deactivates the platform.
    /// </summary>
    public virtual void Deactivate() { }

    /// <summary>
    /// Converts to string.
    /// </summary>
    /// <returns>
    /// A <see cref="System.String" /> that represents this instance.
    /// </returns>
    public override string ToString() => Name;
}

/// <summary>
/// UnknownPlatform
/// </summary>
public class UnknownPlatform : Platform
{
    public static readonly Platform This = new UnknownPlatform();
    UnknownPlatform() : base("UK", "Unknown") { }
}

/// <summary>
/// PlatformX
/// </summary>
public static class PlatformX
{
    /// <summary>
    /// The platform OS.
    /// </summary>
    public enum OS { Windows, OSX, Linux, Android }

    /// <summary>
    /// Gets the platform os.
    /// </summary>
    public static readonly OS PlatformOS = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? OS.Windows
        : RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? OS.OSX
        : RuntimeInformation.OSDescription.StartsWith("android-") ? OS.Android
        : RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ? OS.Linux
        : throw new ArgumentOutOfRangeException(nameof(RuntimeInformation.IsOSPlatform), RuntimeInformation.OSDescription);

    /// <summary>
    /// Gets the platform startups.
    /// </summary>
    public static readonly HashSet<Platform> Platforms = [UnknownPlatform.This];

    /// <summary>
    /// Determines if in a test host.
    /// </summary>
    public static readonly bool InTestHost = AppDomain.CurrentDomain.GetAssemblies().Any(x => x.FullName.StartsWith("testhost,"));

    /// <summary>
    /// Gets the application Path
    /// </summary>
    public static readonly string ApplicationPath = AppDomain.CurrentDomain.BaseDirectory;

    /// <summary>
    /// Gets the platform startups.
    /// </summary>
    public static Dictionary<object, object> Options = DecodeOptions();

    /// <summary>
    /// Gets or sets the current platform.
    /// </summary>
    public static Platform Current = Activate(InTestHost ? TestPlatform.This : UnknownPlatform.This);

    /// <summary>
    /// Activates a platform.
    /// </summary>
    public static Platform Activate(Platform platform)
    {
        if (platform == null || !platform.Enabled) platform = UnknownPlatform.This;
        Platforms.Add(platform);
        var current = Current;
        if (current != platform)
        {
            current?.Deactivate();
            platform?.Activate();
            Current = platform;
        }
        return platform;
    }

    /// <summary>
    /// The platform stats.
    /// </summary>
    public class Stats
    {
        static readonly bool HighRes = Stopwatch.IsHighResolution;
        static readonly double HighFrequency = 1000.0 / Stopwatch.Frequency;
        static readonly double LowFrequency = 1000.0 / TimeSpan.TicksPerSecond;
        static readonly bool UseHrt = false;
        public static bool UsingHighResolutionTiming => UseHrt && HighRes && !Unix;
        public static long TickCount => (long)Ticks;
        public static double Ticks => UseHrt && HighRes && !Unix ? Stopwatch.GetTimestamp() * HighFrequency : DateTime.UtcNow.Ticks * LowFrequency;
        public static readonly bool Is64Bit = Environment.Is64BitProcess;
        public static bool MultiProcessor { get; private set; }
        public static int ProcessorCount { get; private set; }
        public static bool Unix { get; private set; }
        public static bool VR { get; private set; }
    }

    /// <summary>
    /// Creates the matcher.
    /// </summary>
    /// <param name="searchPattern">The searchPattern.</param>
    /// <returns></returns>
    public static Func<string, bool> CreateMatcher(string searchPattern)
    {
        if (string.IsNullOrEmpty(searchPattern)) return x => true;
        var wildcardCount = searchPattern.Count(x => x.Equals('*'));
        if (wildcardCount <= 0) return x => x.Equals(searchPattern, StringComparison.CurrentCultureIgnoreCase);
        else if (wildcardCount == 1)
        {
            var newPattern = searchPattern.Replace("*", "");
            if (searchPattern.StartsWith("*")) return x => x.EndsWith(newPattern, StringComparison.CurrentCultureIgnoreCase);
            else if (searchPattern.EndsWith("*")) return x => x.StartsWith(newPattern, StringComparison.CurrentCultureIgnoreCase);
        }
        var regexPattern = $"^{Regex.Escape(searchPattern).Replace("\\*", ".*")}$";
        return x =>
        {
            try { return Regex.IsMatch(x, regexPattern); }
            catch { return false; }
        };
    }

    public static Dictionary<object, object> DecodeOptions()
    {
        var path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".gamex");
        return File.Exists(path)
            ? (Dictionary<object, object>)new DeserializerBuilder().WithNamingConvention(UnderscoredNamingConvention.Instance).Build()
                .Deserialize(File.ReadAllText(path))
            : default;
    }

    public static string DecodePath(string path, string rootPath = null) =>
        path.StartsWith("%Path%", StringComparison.OrdinalIgnoreCase) ? $"{rootPath}{path[6..]}"
        : path.StartsWith("%AppPath%", StringComparison.OrdinalIgnoreCase) ? $"{ApplicationPath}{path[9..]}"
        : path.StartsWith("%AppData%", StringComparison.OrdinalIgnoreCase) ? $"{Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData)}{path[9..]}"
        : path.StartsWith("%LocalAppData%", StringComparison.OrdinalIgnoreCase) ? $"{Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData)}{path[14..]}"
        : path;
}

#endregion

#region TestPlatform

public interface ITestGfx : IOpenGfx { }

public class TestGfx(PakFile source) : ITestGfx
{
    readonly PakFile _source = source;
    public object Source => _source;
    public Task<T> LoadFileObject<T>(object path) => throw new NotSupportedException();
    public void PreloadTexture(object path) => throw new NotSupportedException();
    public void PreloadObject(object path) => throw new NotSupportedException();
}

public interface ITestSfx : IOpenSfx { }

public class TestSfx(PakFile source) : ITestSfx
{
    readonly PakFile _source = source;
    public object Source => _source;
}

/// <summary>
/// TestPlatform
/// </summary>
public class TestPlatform : Platform
{
    public static readonly Platform This = new TestPlatform();
    TestPlatform() : base("TT", "Test")
    {
        GfxFactory = source => new TestGfx(source);
        SfxFactory = source => new TestSfx(source);
    }
}

#endregion
